Fix/Fix new order triggers when creating/moving orders into draft or from draft to valid statuses

wavvves opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR resolves issues with new order hook triggers during transitions to and from draft status as well as creating a new order with a draft status, using CPT and HPOS. It fixes the following scenarios:

On CPT

- Placing the order, the new order hook is not triggered ❌
- In the admin, creating a new draft order triggers the new order hook with checkout-draft status ❌
- In the admin, changing an order from draft to completed doesn't trigger the new order hook ❌

On HPOS

- In the admin, creating a new draft order triggers the new order hook with checkout-draft status ❌


Closes #48342 
Closes #48080
Closes #48343 

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

### Testing preparation:
1. Install the Debug Log Manager and Code Snippets extensions and add the following snippet:

```php
add_action( 'woocommerce_new_order', 'any', 20, 1 );
function any( $order_id ) {
	$order = wc_get_order( $order_id );
	error_log( 'Order ID: ' . $order_id . ' with status: ' . $order->get_status() );
}
```
2. Keep a tab open with `/wp-admin/tools.php?page=debug-log-manager` and be sure to clear the log before starting the test.

### Testing Steps
1. Clear the error log and go to WooCommerce > Orders and click "Add order"
2. Edit the new order and select "**Draft**" from the status dropdown.
3. Click "Create" and verify that "Created Order ID: xxx with status.. " IS NOT shown on the log manager.
4. Change the Order status again to `Completed` click `Update` and verify that "Created Order ID: xxx with status.. " IS shown on the log manager.
5. Clear the log
6. Make an order using the Checkout Blocks
7. Verify that "Created Order ID: xxx with status.. " IS shown on the log manager.
8. Go to `Admin` > `WooCommerce` > `Advanced` > `Features` and switch between Wordpress posts storage/High-performance order storage (depending on the previous choice). You might have to sync orders before doing so.
<img width="400" alt="image" src="https://github.com/woocommerce/woocommerce/assets/17236129/9b65a3d5-60e7-4769-acba-6a1293d88dfe">

9. Repeat steps 1-8

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

wavvves committed:
Prevent new order hook if moving or creating order status into a non-triggering status (drafts&failed refunded etc) when using CPT for orders.

wavvves committed:
Removed rogue error_log()

wavvves committed:
Lint

wavvves committed:
Prevent new order hook if moving or creating order status into a non-triggering status (drafts&failed refunded etc) when using HPOS for orders.

wavvves committed:
Merge branch 'trunk' into fix/48342-in-the-admin-changing-an-order-from-draft-to-completed-doesnt-trigger-anything-when-using-cpt

wavvves committed:
Added changelog

wavvves committed:
Fixed wrong if simplification

wavvves committed:
Unit tests for update methods

wavvves committed:
Merge branch 'trunk' into fix/48342-in-the-admin-changing-an-order-from-draft-to-completed-doesnt-trigger-anything-when-using-cpt

mikejolley commented on plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php:
```diff
@@ -193,18 +193,34 @@ public function update( &$order ) {
-		if ( $new_status !== $previous_status && in_array( $previous_status, array( 'new', 'auto-draft', 'draft', 'checkout-draft' ), true ) ) {
+		// We need to remove the wc- prefix from the status for comparison and proper evaluation of new vs updated orders.
+		if ( strpos( $previous_status, 'wc-' ) === 0 ) {
+			$previous_status = substr( $previous_status, 3 );
```
nit: `current_status` might be more understandable. We're looking at current status from the DB before the order object is updated.

wavvves replied:
done in https://github.com/woocommerce/woocommerce/pull/49098/commits/146b14297abba200d644cfda31bd59a52ec059e0 👍🏼

mikejolley commented on plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php:
```diff
@@ -193,18 +193,34 @@ public function update( &$order ) {
+			$previous_status = substr( $previous_status, 3 );
+		}
+
+		if ( strpos( $new_status, 'wc-' ) === 0 ) {
```
The abstract order class has some code which removes prefix in the `set_status`  method. How about creating a utility method that this can share?

wavvves replied:
Good call, added in https://github.com/woocommerce/woocommerce/commit/8b50abaf6bd7e9f48540559799752e562f75b5db and replaced a few other places in the code that was doing the same thing 👍🏼

mikejolley commented on plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php:
```diff
@@ -193,18 +193,34 @@ public function update( &$order ) {
+			$new_status = substr( $new_status, 3 );
+		}
+
+		$nonTriggeringOrderStatuses = array( 'new', 'auto-draft', 'draft', 'checkout-draft', 'refunded', 'failed', 'cancelled');
```
Check your IDE is setup to show linting rules. This violates naming (snake_case is needed on the PHP side).

wavvves replied:
It is highlighting, I don't know what glitch got into it or my brain 🫠
Changed in https://github.com/woocommerce/woocommerce/pull/49098/commits/3f28c6b664885f7ee475803df85b1766eca56dc7

mikejolley commented on plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php:
```diff
@@ -193,18 +193,34 @@ public function update( &$order ) {
+		if (
+			$new_status !== $previous_status
+			&& ! in_array( $new_status, $nonTriggeringOrderStatuses, true )
+			&& in_array( $previous_status, $nonTriggeringOrderStatuses, true )
```
Does this work reliably if creating a new order from scratch? The `$previous_status` comes from the database, so if the order does not exist in the database yet will this break?

wavvves replied:
We can implement a similar solution to what you suggested with HPOS. In this case, the `$previous_status` will be `false` if it fails to retrieve it from the database. By setting it to `new,` we can correctly trigger the hook if it lands outside the non-triggering statuses. Do you agree?

```php
// Also grab the current status so we can compare.
$previous_status = get_post_status( $order->get_id() );
// If the order doesn't exist in the DB, we will consider it as new.
if ( ! $previous_status ) {
    $previous_status = 'new';
}
```

I added this in https://github.com/woocommerce/woocommerce/pull/49098/commits/b733fa699ea423b3262b80d818519dc7d3dbd7b4 Let me know what you think

mikejolley commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -2621,8 +2621,15 @@ public function update( &$order ) {
+			! empty( $changes['status'] )
+			&& $changes['status'] !== $previous_status
+			&& ! in_array( $changes['status'], $nonTriggeringOrderStatuses, true )
+			&& in_array( $previous_status, $nonTriggeringOrderStatuses, true )
```
Looking at the code, in this context, a brand new unsaved order would have $previous_status set to `status`?

```
$previous_status = ArrayUtil::get_value_or_default( $order->get_data(), 'status' );
```

Thus this would break.

mikejolley replied:
We could default the status to `new` is the order is not yet saved in DB.

wavvves replied:
The second arg is `$key`, this would actually return `null` as `$previous_status`. But what you say makes sense, an order that exists only in memory going from `null` status into one of the triggering status should trigger the new order hook!

Changed in https://github.com/woocommerce/woocommerce/pull/49098/commits/f0b4303eea5c00f9e47cf28acc54af531c40fb6f

mikejolley commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -2621,8 +2621,15 @@ public function update( &$order ) {
 
-		// For backwards compatibility, moving a draft order to a valid status triggers the 'woocommerce_new_order' hook.
-		if ( ! empty( $changes['status'] ) && in_array( $previous_status, array( 'new', 'auto-draft', 'draft', 'checkout-draft' ), true ) ) {
+		$nonTriggeringOrderStatuses = array( 'new', 'auto-draft', 'draft', 'checkout-draft', 'refunded', 'failed', 'cancelled');
```
More snake_case issues.

wavvves replied:
https://github.com/woocommerce/woocommerce/pull/49098/commits/3f28c6b664885f7ee475803df85b1766eca56dc7

mikejolley commented on plugins/woocommerce/tests/php/includes/data-stores/class-wc-order-data-store-cpt-test.php:
```diff
@@ -466,4 +466,50 @@ public function test_create_draft_order_doesnt_trigger_hook() {
+
+		$order_data_store_cpt = new WC_Order_Data_Store_CPT();
+
+		$order = new WC_Order();
```
I would add another test case here to create a new order, set status to `processing` immediately, and update. It should trigger `woocommerce_new_order` but I suspect it will not until my other feedback is resolved.

wavvves replied:
Like this https://github.com/woocommerce/woocommerce/pull/49098/commits/4848866e2089c2fd58af3f4e5c937150b996bcff ?

mikejolley replied:
Yes 👍🏻

mikejolley reviewed (changes requested):
Added some questions and feedback r/e linting and new orders created manually without first creating a draft.

wavvves committed:
Renamed var

wavvves committed:
Renamed vars to snake_case.

wavvves committed:
Introduced remove_status_prefix() utility function.

wavvves committed:
Merge branch 'trunk' into fix/48342-in-the-admin-changing-an-order-from-draft-to-completed-doesnt-trigger-anything-when-using-cpt

# Conflicts:
#	plugins/woocommerce/includes/wc-order-functions.php

wavvves committed:
Added default 'new' status to an Order that doesn't exist yet in the database for the update method in HPOS

wavvves committed:
Added default 'new' status to an Order that doesn't exist yet in the database for the update method in CPT

wavvves committed:
Linting.

wavvves committed:
Missing class import.

wavvves committed:
Merge branch 'trunk' into fix/48342-in-the-admin-changing-an-order-from-draft-to-completed-doesnt-trigger-anything-when-using-cpt

wavvves committed:
Added a new test for updating new processing Orders.

wavvves committed:
Linting.

wavvves committed:
Tests fix.

wavvves committed:
Tests fix.

wavvves committed:
Merge branch 'trunk' into fix/48342-in-the-admin-changing-an-order-from-draft-to-completed-doesnt-trigger-anything-when-using-cpt

mikejolley reviewed (approved):
Changes since last review look good. Thanks for taking care of those improvements. Since tests are green I think you're good to go.

wavvves commented:
> Changes since last review look good. Thanks for taking care of those improvements. Since tests are green I think you're good to go.

Thanks Mike! I just wait for a look by Proton before merging 👍🏼 cc/ @vedanshujain

vedanshujain commented on plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php:
```diff
@@ -189,22 +191,37 @@ public function update( &$order ) {
 		// Also grab the current status so we can compare.
 		$previous_status = get_post_status( $order->get_id() );
+		// If the order doesn't exist in the DB, we will consider it as new.
+		if ( ! $previous_status ) {
```
get_post_status can be false when there's a intermittent DB issue, and not necessarily if the post does not exist. A better check for not existing in DB is to check fo `$order->get_id() > 0`

wavvves replied:
could this be `( ! $previous_status && $order->get_id() === 0 )` or is it too redundant? Since we have the bool already, it would be quick to evaluate and save the `get_id()` call.
We're only changing order status in a non-persistent way here to allow the triggering of the hook, as per Mike's comment

wavvves replied:
I added the check on https://github.com/woocommerce/woocommerce/pull/49098/commits/1ab534aabc39613aa8f21e872a0b3087b2b2c0d0 in case you are onboard with it and decide to approve 👍🏼

vedanshujain commented on plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php:
```diff
@@ -189,22 +191,37 @@ public function update( &$order ) {
+		$previous_status = OrderUtil::remove_status_prefix( $previous_status );
+		$current_status  = OrderUtil::remove_status_prefix( $current_status );
+
+		$non_triggering_order_statuses = array( 'new', 'auto-draft', 'draft', 'checkout-draft', 'refunded', 'failed', 'cancelled' );
```
can you add the justification for adding `'refunded', 'failed', 'cancelled'` here, I don't think we were doing this before?

wavvves replied:
Removing these as, after discussion, the conclusion was each order ID should trigger one new order event 👍🏼

wavvves replied:
https://github.com/woocommerce/woocommerce/pull/49098/commits/1175b38586a630bedc6bd969dde03e57adf3a610

vedanshujain reviewed (commented):
Added a questions around new status before we can merge

wavvves committed:
Removed statuses from triggering the new order hook.

wavvves committed:
Merge branch 'trunk' into fix/48342-in-the-admin-changing-an-order-from-draft-to-completed-doesnt-trigger-anything-when-using-cpt

wavvves committed:
tweak

wavvves committed:
Test tweak

wavvves committed:
Test tweak

wavvves committed:
Test tweak

wavvves committed:
Test tweak

wavvves committed:
Test tweak

wavvves committed:
Tweak to rule out db querying failure.

wavvves committed:
Merge branch 'trunk' into fix/48342-in-the-admin-changing-an-order-from-draft-to-completed-doesnt-trigger-anything-when-using-cpt

vedanshujain reviewed (approved):
LGTM! Code looks good and verified hooks with sync enabled on both HPOS and posts data store. Note that I also updated the snippet in testing instruction to handle both posts and HPOS.
